tests/unit/crawlers/_http/test_http_crawler.py

```diff
@@ -410,3 +417,71 @@ def mark_request_execution(request: Request) -> Response:  # noqa: ARG001 # Unus
     await crawler.run([Request.from_url(url=test_url)])
 
     assert execution_order == ['pre-navigation-hook 1', 'pre-navigation-hook 2', 'request', 'final handler']
+
+
+@pytest.mark.parametrize(
+    'http_client_class',
+    [CurlImpersonateHttpClient, HttpxHttpClient],
+    ids=['curl', 'httpx'],
+)
+async def test_isolation_cookies(http_client_class: type[BaseHttpClient], httpbin: URL) -> None:
+    http_client = http_client_class()
+    sessions_id: list[str] = []
+    sessions_cookies: dict[str, dict[str, str]] = {}
+    response_cookies: dict[str, dict[str, str]] = {}
+
+    crawler = HttpCrawler(
+        session_pool=SessionPool(max_pool_size=1),
+        http_client=http_client,
+    )
+
+    @crawler.router.default_handler
+    async def handler(context: HttpCrawlingContext) -> None:
+        if context.session:
+            sessions_id.append(context.session.id)
+
+        # Add to the queue the request that will be made by the session with the cookie
+        await context.add_requests(
+            [
+                Request.from_url(str(httpbin.with_path('/cookies')), unique_key='1', label='cookies'),
+            ]
+        )
+
+    @crawler.router.handler('cookies')
+    async def cookies_handler(context: HttpCrawlingContext) -> None:
+        if context.session:
+            if context.request.unique_key == '2':
+                sessions_id.append(context.session.id)
+            sessions_cookies[context.session.id] = context.session.cookies
+            response_data = json.loads(context.http_response.read())
+            response_cookies[context.session.id] = response_data.get('cookies')
+            context.session.retire()
+
+        # The session with the cookie is retire. The next request should be made by a session without a cookie
+        if context.request.unique_key == '1':
+            await context.add_requests(
+                [
+                    Request.from_url(str(httpbin.with_path('/cookies')), unique_key='2', label='cookies'),
+                ]
+            )
```

Comment:
It is kinda hard to understand how the two request handlers interact - any chance they could be merged?

Reply:
Agreed, I had a hard time understanding this test as well :smile:.